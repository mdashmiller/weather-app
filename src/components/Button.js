import styled from "styled-components"

const Button = styled.button`
	height: 37px;
    width: 50px;
    margin-left: 8px;
    border-radius: 4px;
    border: 2px solid #c3c3c3;
    color: #c3c3c3;
    background: #4c7d98;
    padding: 0;
    font-size: 18px;
    font-family: inherit;
    font-weight: 900;
    cursor: pointer;

    	&:hover {
    		background: #c3c3c3;
    		color: #4c7d98;
    		border: none;
    		outline: none;
    	} 

    	&:active,
    	:focus,
    	:visited {
    		outline: none;
    	}
`

export default Button
