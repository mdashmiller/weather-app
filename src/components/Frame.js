import styled from "styled-components"

const Frame = styled.div`
	width: 300px;
	margin: 40px auto 30px;
	text-align: center;
	font-family: 'Varela Round', Arial, sans-serif;
`

export default Frame
