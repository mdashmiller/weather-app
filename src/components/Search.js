import styled, { css } from "styled-components"

const Search = styled.input`
	background: #eee;
	border: 1px solid #c3c3c3;
	border-radius: 10px;
	margin: 200px auto 20px;
	width: 200px;
	height: 34px;
	padding-left: 8px;
	font-family: 'Font Awesome 5 Free', 'Varela Round', Arial, sans-serif;
	font-weight: 900;
	font-size: 18px;

	&::placeholder {
		color: #6a6a6a;
	}

	${props => props.short && css`
	    width: 114px;
	    padding-left: 5px;
	    margin: 0;
	  `}

	${props => props.landingPage && css`
		display: block;
	    width: 172px;
	   	margin: 26px auto;
		text-align: center;
		padding: 0;
		cursor: pointer;  
	  `}

	${props => props.userLocation && css`
		border: 3px solid #f46403;
			
			&:hover {
				background: #f46403;
				
				&::placeholder {
					color: #eee;
				}
			}	
	  `}

	${props => props.userZip && css`
		border: 3px solid #6a6a6a;

			&:hover {
				background: #6a6a6a;
				
				&::placeholder {
					color: #eee;
				}
			} 
	  `}
`

export default Search
