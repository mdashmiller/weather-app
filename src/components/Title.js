import React from 'react'
import { Link } from 'react-router-dom'

const Title = () =>
	<Link to='/'>
		<h1 id="landing-title" className="title-gold">
			Weather
			<span className="title-grey">
				Now
			</span>
		</h1>
	</Link>

export default Title
