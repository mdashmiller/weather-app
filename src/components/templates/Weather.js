import React, { Component } from 'react'
import WeatherIcon from '../ui/WeatherIcon'
import ThermoIcon from '../ui/ThermoIcon'
import PropTypes from 'prop-types'

class Weather extends Component {

	state = {
		day: null,
		place: '',
		temp: '',
		description: '', 
		codition: null
	}

	// component methods

	setWeatherInfo = result => {
		// takes results from API call and
		// routes data to the proper channels
		this.setTemp(result.main.temp)
		this.setDescription(result.weather[0].description)
		this.setPlace(result.name)
		this.setCondition(result.weather[0].id)
		this.dayOrNight(result.dt, result.sys.sunrise, result.sys.sunset)
	}

	setTemp = tempK => {
		// sets this.state.temp to the result
		// obtained from Open Weather Map
		const tempF =  this.convertToFahrenheit(tempK)
		this.setState( {temp: tempF} )
	}

	convertToFahrenheit = k => {
		// takes a temperature in Kelvin
		// and returns it in Fahrenheit
		return (
			(1.8 * (k - 273)) + 32
		).toFixed(1)
	}

	setDescription = weather =>
		// sets this.state.description to the result
		// obtained from Open Weather Map
		this.setState( {description: weather })

	setPlace = place =>
		// sets this.state.name to the result
		// obtained from Open Weather Map
		this.setState( {place} )

	setCondition = id =>
		// sets this.state.condition to the result
		// obtained from Open Weather Map
		this.setState( {condition: id} )

	dayOrNight = (time, sunrise, sunset) => {
		// determines if it is day or night
		// and sets state accordingly
		if (time >= sunrise && time < sunset) {
			this.setState({ day: true })
		} else {
			this.setState({ day: false })
		}
	}

	// lifecycle hooks

	componentDidMount () {
		this.setWeatherInfo(this.props.result)
	}

	render () {
		const {
			day,
			place,
			temp,
			description,
			condition
		} = this.state
		return (
			<div>
				<h1 className="title-gold">
					{place}
				</h1>
				<h2>{temp} F</h2>
				<h3>{description}</h3>
				<WeatherIcon 
					code={condition} 
					day={day}
				/>
				<ThermoIcon temp={temp} />
			</div>
		)
	}
}

Weather.propTypes = {
	result: PropTypes.shape({
		name: PropTypes.string,
		main: PropTypes.shape({
			temp: PropTypes.number
		}),
		weather: PropTypes.arrayOf(
			PropTypes.shape({
				description: PropTypes.string,
				id: PropTypes.number
			})
		),
		dt: PropTypes.number,
		sys: PropTypes.shape({
			sunrise: PropTypes.number,
			sunset: PropTypes.number
		})
	})
}

export default Weather
