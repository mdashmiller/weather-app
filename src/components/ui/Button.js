import styled from "styled-components"

const Button = styled.button`
	height: 37px;
    width: 50px;
    margin-left: 8px;
    border-radius: 4px;
    border: 2px solid #eee;
    color: #eee;
    background: #52606b;
    padding: 0;
    font-size: 18px;
    font-family: inherit;
    font-weight: 900;
    cursor: pointer;

    	&:hover {
    		background: #f2b632;
    		border: none;
    		outline: none;
    	} 

    	&:active,
    	:focus,
    	:visited {
    		outline: none;
    	}
`

export default Button
