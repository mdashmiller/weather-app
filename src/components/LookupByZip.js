import React, { Component } from 'react'
import config from '../config'
import { Link } from 'react-router-dom'
import Frame from './Frame'
import Search from './Search'
import Button from './Button'
import Weather from './Weather'

// Open Weather Map API url details
const PATH_BASE = 'http://api.openweathermap.org/data/2.5/weather?'
const KEY = config.key

class LookupByZip extends Component {

	state = {
		zip: '',
		result: {},
		error: null,
		searchClicked: false
	}

	// component methods

	handleZip = e => 
		// updates this.state.zip according
		// to user input
		this.setState({ zip: e.target.value })

	handleKeyPress = e => {
		// handles user keypresses in the lookup
		// by zipcode input and limits the quantity 
		// and quality of chars the user can enter
		const numOfChars = e.target.value.length	
		const char = e.key || e.keyCode
		// submit with enter key
		if (char === 13 || char === 'Enter' ) {
			this.getWeather()
		}
		// allows a max of 5 chars
		if (numOfChars < 5) {
			// input will only accept [0 - 9]
			if (char >= 48 && char <= 57) {
				return
			} else if (char >= 0 && char <= 9) {
				return
			} else {
				e.preventDefault()
			}
		} else {
			// backspace, left, right, enter, tab, home, end, 
			// and return keys can be used regardless of 
			// number of chars enterd
			if (char === 8 || char === 'Backspace'
				|| char === 37 || char === 'ArrowLeft' || char === 'Left'
				|| char === 39 || char === 'ArrowRight' || char === 'Right'
				|| char === 13 || char === 'Enter'
				|| char === 9 || char === 'Tab'
				|| char === 36 || char === 'Home'
				|| char === 35 || char === 'End') {
				return
			} else {
				e.preventDefault()
			}
		}
	}

	handleClick = () => {
		this.startSpinner()
		this.getWeather()
	}

	startSpinner = () => 
		this.setState({ searchClicked: true })

	getWeather = () => {
		// call weather API with zipcode
		const ZIP = `zip=${this.state.zip}`	
		const urlZip = `${PATH_BASE}${ZIP}&APPID=${KEY}`
		fetch(urlZip)
			.then(response => response.json())
			.then(result => this.setWeatherInfo(result))
			.catch(error => this.setState({ error }))
	}

	setWeatherInfo = result =>
		// set result from API call
		// and reset zip in state
		this.setState({
			result,
			zip: '',
			searchClicked: false
		})

	clearWeather = () => 
		// resets state to prepare for
		// a new API call
		this.setState({
			result: {},
			error: null
		})
	
	render () {
		const {
			result,
			zip,
			searchClicked
		} = this.state
		if (!result.name) {
			return (
				<Frame>
					<h1 id="landing-title" className="title-gold">
						Weather
						<span className="title-grey">
							Now
						</span>
					</h1>
					<Search
						type="text"
						placeholder="Enter US ZIP"
						value={zip}
						autoFocus
						short
						onChange={this.handleZip}
						onKeyPress={this.handleKeyPress}
						id={ result.cod ? 'error-box' : undefined }
					/>
					{	!searchClicked 
							? (
								<Button onClick={this.handleClick}>
									Go
								</Button>
							) : (
								<i className="fas fa-spinner fa-2x zip-spinner">
								</i>
							)
					}
					{	result.cod &&
							<h2 id="error">Hm, that doesn't seem to be a valid US zip. Try again?</h2>
					}
				</Frame>
			)
		} else {
			return (
				<Frame>
					<Weather
						result={result}
					/>
					<Search
						type="text"
						placeholder="&#xf002; Change Location"
						value={zip}
						onClick={this.clearWeather}
						lookUp
					/>
					<Link to='/lookup-by-geoloc'>			    	
						<Search
							autoFocus
							type="text"
							placeholder="&#xf3c5;  Use My Location"
							displayPage
							userLocation
						/>
					</Link>
				</Frame>
			)
		}
	}
}

export default LookupByZip
