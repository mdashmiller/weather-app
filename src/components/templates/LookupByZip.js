import React, { Component } from 'react'
import config from '../../config'
import { Link } from 'react-router-dom'
import Weather from './Weather'
import ErrorMsg from './ErrorMsg'
import Frame from '../ui/Frame'
import Search from '../ui/Search'
import Button from '../ui/Button'
import Title from '../ui/Title'

// Open Weather Map API url details
const PATH_BASE = 'http://api.openweathermap.org/data/2.5/weather?'
const KEY = config.key

class LookupByZip extends Component {

	state = {
		zip: '',
		result: {},
		error: null,
		searchClicked: false
	}

	// component methods

	handleZip = e => 
		// updates this.state.zip according
		// to user input
		this.setState({ zip: e.target.value })

	handleKeyPress = e => {
		// handles user keypresses in the lookup
		// by zipcode input and limits the quantity 
		// and quality of chars the user can enter
		const numOfChars = e.target.value.length	
		const char = e.key || e.keyCode
		const selection = document.getSelection().toString()
		// submit with enter key
		if (char === 13 || char === 'Enter' ) {
			this.getWeather()
		}
		// allows a max of 5 chars and allows user
		// to replace any chars that are highlighted
		if (numOfChars < 5 || selection !== '') {
			// input will only accept [0 - 9]
			if (char >= 48 && char <= 57) {
				return
			} else if (char >= 0 && char <= 9) {
				return
			} else {
				e.preventDefault()
			}
		} else {
			// backspace, left, right, enter, tab, home, end, 
			// and return keys can be used regardless of 
			// number of chars enterd
			if (char === 8 || char === 'Backspace'
				|| char === 37 || char === 'ArrowLeft' || char === 'Left'
				|| char === 39 || char === 'ArrowRight' || char === 'Right'
				|| char === 13 || char === 'Enter'
				|| char === 9 || char === 'Tab'
				|| char === 36 || char === 'Home'
				|| char === 35 || char === 'End') {
				return
			} else {
				e.preventDefault()
			}
		}
	}

	clearError = () => {
		// if there was a fetch error, this will 
		// clear the API fail message and reset 
		// the form when user
		// goes to enter a new zip
		this.setState({
			error: null,
			searchClicked: false
		})
	}

	handleClick = () => {
		// handles the click event for the
		// search-by-zip submit button
		this.startSpinner()
		this.getWeather()
	}

	startSpinner = () => 
		// sets state so that a conditional
		// loading spinner will be rendered
		this.setState({ searchClicked: true })

	getWeather = () => {
		// call weather API with zipcode
		const ZIP = `zip=${this.state.zip}`	
		const urlZip = `${PATH_BASE}${ZIP}&APPID=${KEY}`
		fetch(urlZip)
			.then(response => response.json())
			.then(result => this.setWeatherInfo(result))
			.catch(error => this.setState({ error }))
	}

	setWeatherInfo = result => {
		result.name ? (
			// set result from API call
			// and reset zip in state on 
			// a successful call
			this.setState({
				result,
				zip: '',
				searchClicked: false
			})
		) : (
			// don't reset the zip if
			// a bad zip was entered
			this.setState({
				result,
				searchClicked: false
			})
		)
	}
		
	clearWeather = () => 
		// resets state to prepare for
		// a new API call
		this.setState({
			result: {},
			error: null
		})
	
	render () {
		const {
			result,
			zip,
			error,
			searchClicked
		} = this.state
		if (!result.name) {
			return (
				<Frame>
					<Title />
					<Search
						type="text"
						placeholder="Enter US ZIP"
						value={zip}
						autoFocus
						short
						onChange={this.handleZip}
						onKeyPress={this.handleKeyPress}
						onClick={this.clearError}
						id={ result.cod ? 'error-box' : undefined }
					/>
					{
						searchClicked && !error
							? (
								<i className="fas fa-spinner fa-2x zip-spinner">
								</i>	
							) : (
								<Button onClick={this.handleClick}>
									Go
								</Button>
							)
					}
					{
						result.cod &&
							<div id="error">
								<h2>Hm, that doesn't seem to be a valid US zip.</h2>
								<h2>Try again?</h2>
							</div>
					}
					{
						error &&
							<ErrorMsg />
					}
				</Frame>
			)
		} else {
			return (
				<Frame>
					<Weather
						result={result}
					/>
					<Search
						type="text"
						placeholder="&#xf002; Change Location"
						value={zip}
						onClick={this.clearWeather}
						lookUp
					/>
					<Link to='/lookup-by-geoloc'>			    	
						<Search
							autoFocus
							type="text"
							placeholder="&#xf3c5;  Use My Location"
							displayPage
							userLocation
						/>
					</Link>
				</Frame>
			)
		}
	}
}

export default LookupByZip
