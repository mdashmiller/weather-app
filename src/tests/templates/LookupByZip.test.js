import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import Enzyme, { shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import { BrowserRouter, Link } from 'react-router-dom'

import LookupByZip from '../../components/templates/LookupByZip'
import Frame from '../../components/ui/Frame'
import Title from '../../components/ui/Title'
import Search from '../../components/ui/Search'
import Button from '../../components/ui/Button'
import Weather from '../../components/templates/Weather'

Enzyme.configure({ adapter: new Adapter() })

describe('<LookupByZip /> basic rendering', () => {

	it('renders without crashing', () => {
	  	const div = document.createElement('div')
	  	ReactDOM.render(
	  		<BrowserRouter>
	  			<LookupByZip />
	  		</BrowserRouter>,
	  	div)
	  	ReactDOM.unmountComponentAtNode(div)
	})

	test('has a valid snapshot', () => {
		const component = renderer.create(
			<BrowserRouter>
				<LookupByZip />
			</BrowserRouter>
		)
		let tree = component.toJSON()
		expect(tree).toMatchSnapshot()
	})

})

describe('<LookupByZip /> rendering when there is no weather data', () => {

	let wrapper

	beforeEach(() => {
		wrapper = shallow(<LookupByZip />)
	})

	it('renders 1 <Frame> component', () => {
		expect(wrapper.find(Frame).length).toBe(1)
	})

	it('renders 1 <Title> component', () => {
		expect(wrapper.find(Title).length).toBe(1)
	})

	it('renders 1 <Search> component', () => {
		expect(wrapper.find(Search).length).toBe(1)
	})

	it('renders 1 <i> element when there is no error and the button has been clicked', () => {
		wrapper.setState({ searchClicked: true })
		expect(wrapper.find('i').length).toBe(1)
	})

	it('renders 1 <Button> component when there is an error or the button has not been clicked', () => {
		wrapper.setState({ error: {} })
		expect(wrapper.find(Button).length).toBe(1)
	})

	it('renders 1 <div> #error when the user enters a bad zip', () => {
		wrapper.setState({
			result: {
				cod: 'error'
			}
		})
		expect(wrapper.find('div').length).toBe(1)
		expect(wrapper.find('div').prop('id')).toBe('error')
	})

	it('renders 2 <h2> elements when the user enters a bad zip', () => {
		wrapper.setState({
			result: {
				cod: 'error'
			}
		})
		expect(wrapper.find('h2').length).toBe(2)
	})

})

describe('<LookupByZip /> rendering when there is weather data', () => {

	let wrapper

	beforeEach(() => {
		wrapper = shallow(<LookupByZip />)
		wrapper.setState({
			result: {
				name: 'Test City'
			}
		})
	})

	it('renders 1 <Frame> component', () => {
		expect(wrapper.find(Frame).length).toBe(1)
	})

	it('renders 1 <Weather> component', () => {
		expect(wrapper.find(Weather).length).toBe(1)
	})

	it('renders 2 <Search> components', () => {
		expect(wrapper.find(Search).length).toBe(2)
	})

})
