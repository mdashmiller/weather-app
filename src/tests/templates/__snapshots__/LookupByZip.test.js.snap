// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LookupByZip /> basic rendering has a valid snapshot 1`] = `
<div
  className="sc-bwzfXH kUBsGr"
>
  <a
    href="/"
    onClick={[Function]}
  >
    <h1
      className="title-gold"
      id="landing-title"
    >
      Weather
      <span
        className="title-grey"
      >
        Now
      </span>
    </h1>
  </a>
  <input
    autoFocus={true}
    className="sc-htpNat dxPalS"
    onChange={[Function]}
    onClick={[Function]}
    onKeyPress={[Function]}
    placeholder="Enter US ZIP"
    type="text"
    value=""
  />
  <button
    className="sc-bxivhb bomdTJ"
    onClick={[Function]}
  >
    Go
  </button>
</div>
`;
